File: SolaceCore/Sources/SolaceCore/ViewModels/TodayViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
public final class TodayViewModel {
    public let currentUser: User
    public private(set) var moodHistory: [MoodEntry] = []
    public private(set) var gratitudeEntries: [GratitudeEntry] = []
    public private(set) var isSubmittingMood = false
    public private(set) var errorMessage: String?

    private let moodService: MoodServicing
    private let journalService: JournalServicing
    @ObservationIgnored
    private nonisolated(unsafe) var moodObservationTask: Task<Void, Never>?
    @ObservationIgnored
    private nonisolated(unsafe) var journalObservationTask: Task<Void, Never>?

    public init(currentUser: User, moodService: MoodServicing, journalService: JournalServicing) {
        self.currentUser = currentUser
        self.moodService = moodService
        self.journalService = journalService
        moodObservationTask = Task { [weak self] in
            guard let self else { return }
            for await history in moodService.observeMoodHistory(for: currentUser.id) {
                self.moodHistory = history
            }
        }
        journalObservationTask = Task { [weak self] in
            guard let self else { return }
            for await entries in journalService.observeGratitudeEntries(for: currentUser.id) {
                self.gratitudeEntries = entries
            }
        }
    }

    deinit {
        moodObservationTask?.cancel()
        journalObservationTask?.cancel()
    }

    public var greeting: String {
        let hour = Calendar.current.component(.hour, from: Date())
        let timeOfDay: String
        switch hour {
        case 0..<12: timeOfDay = "morning"
        case 12..<17: timeOfDay = "afternoon"
        default: timeOfDay = "evening"
        }
        return "Good \(timeOfDay), \(currentUser.displayName)"
    }

    public var hasCheckedInToday: Bool {
        guard let mostRecent = moodHistory.first else { return false }
        return Calendar.current.isDateInToday(mostRecent.createdAt)
    }

    /// Consecutive calendar days (ending today or yesterday — a missed today
    /// doesn't zero out yesterday's progress until tomorrow) with at least
    /// one mood check-in or journal entry.
    public var streak: Int {
        let calendar = Calendar.current
        var days = Set(moodHistory.map { calendar.startOfDay(for: $0.createdAt) })
        days.formUnion(gratitudeEntries.map { calendar.startOfDay(for: $0.createdAt) })
        guard !days.isEmpty else { return 0 }

        var cursor = calendar.startOfDay(for: Date())
        if !days.contains(cursor) {
            guard let yesterday = calendar.date(byAdding: .day, value: -1, to: cursor), days.contains(yesterday) else {
                return 0
            }
            cursor = yesterday
        }

        var streakCount = 0
        while days.contains(cursor) {
            streakCount += 1
            guard let previous = calendar.date(byAdding: .day, value: -1, to: cursor) else { break }
            cursor = previous
        }
        return streakCount
    }

    public func logMood(_ mood: MoodLevel, note: String? = nil) async {
        errorMessage = nil
        isSubmittingMood = true
        defer { isSubmittingMood = false }
        do {
            try await moodService.logMood(mood, note: note, for: currentUser.id)
        } catch {
            errorMessage = "Couldn't save your check-in. Please try again."
        }
    }
}

File: SolaceAndroid/app/src/main/java/com/jonathanalumasa/solace/viewmodel/TodayViewModel.kt
package com.jonathanalumasa.solace.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jonathanalumasa.solace.model.GratitudeEntry
import com.jonathanalumasa.solace.model.MoodEntry
import com.jonathanalumasa.solace.model.MoodLevel
import com.jonathanalumasa.solace.model.User
import com.jonathanalumasa.solace.service.JournalService
import com.jonathanalumasa.solace.service.MoodService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Calendar
import java.util.Date

/** Mirrors `SolaceCore.TodayViewModel`. */
class TodayViewModel(
    val currentUser: User,
    private val moodService: MoodService,
    private val journalService: JournalService
) : ViewModel() {

    private val _moodHistory = MutableStateFlow<List<MoodEntry>>(emptyList())
    val moodHistory: StateFlow<List<MoodEntry>> = _moodHistory.asStateFlow()

    private val _gratitudeEntries = MutableStateFlow<List<GratitudeEntry>>(emptyList())
    val gratitudeEntries: StateFlow<List<GratitudeEntry>> = _gratitudeEntries.asStateFlow()

    private val _isSubmittingMood = MutableStateFlow(false)
    val isSubmittingMood: StateFlow<Boolean> = _isSubmittingMood.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    init {
        viewModelScope.launch {
            moodService.observeMoodHistory(currentUser.id).collect { _moodHistory.value = it }
        }
        viewModelScope.launch {
            journalService.observeGratitudeEntries(currentUser.id).collect {
                _gratitudeEntries.value = it
            }
        }
    }

    fun clearError() {
        _errorMessage.value = null
    }

    val greeting: String
        get() {
            val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
            val timeOfDay = when (hour) {
                in 0..11 -> "morning"
                in 12..16 -> "afternoon"
                else -> "evening"
            }
            return "Good $timeOfDay, ${currentUser.displayName}"
        }

    val hasCheckedInToday: Boolean
        get() {
            val mostRecent = _moodHistory.value.firstOrNull() ?: return false
            return isSameDay(mostRecent.createdAt, Date())
        }

    /**
     * Consecutive calendar days (ending today or yesterday — a missed today
     * doesn't zero out yesterday's progress until tomorrow) with at least one
     * mood check-in or journal entry.
     */
    val streak: Int
        get() {
            val days = buildSet {
                _moodHistory.value.forEach { add(startOfDay(it.createdAt)) }
                _gratitudeEntries.value.forEach { add(startOfDay(it.createdAt)) }
            }
            if (days.isEmpty()) return 0

            var cursor = startOfDay(Date())
            if (!days.contains(cursor)) {
                val yesterday = addDays(cursor, -1)
                if (!days.contains(yesterday)) return 0
                cursor = yesterday
            }

            var count = 0
            while (days.contains(cursor)) {
                count += 1
                cursor = addDays(cursor, -1)
            }
            return count
        }

    fun logMood(mood: MoodLevel, note: String? = null) {
        viewModelScope.launch {
            _errorMessage.value = null
            _isSubmittingMood.value = true
            try {
                moodService.logMood(mood, note, currentUser.id)
            } catch (error: Exception) {
                _errorMessage.value = "Couldn't save your check-in. Please try again."
            } finally {
                _isSubmittingMood.value = false
            }
        }
    }

    private fun startOfDay(date: Date): Long = Calendar.getInstance().apply {
        time = date
        set(Calendar.HOUR_OF_DAY, 0)
        set(Calendar.MINUTE, 0)
        set(Calendar.SECOND, 0)
        set(Calendar.MILLISECOND, 0)
    }.timeInMillis

    private fun addDays(startOfDayMillis: Long, days: Int): Long =
        Calendar.getInstance().apply {
            timeInMillis = startOfDayMillis
            add(Calendar.DAY_OF_YEAR, days)
        }.timeInMillis

    private fun isSameDay(a: Date, b: Date): Boolean = startOfDay(a) == startOfDay(b)
}
